### PR TITLE
Write Fraxlend (v1 & v2) Collateral and Debt Token Adaptors

### PR DESCRIPTION
**Pull request by 0xEinCodes**
## Core changes:

This PR encompasses the development and unit testing of the FRAX collateral and debt token adaptors. 

There are a three different phases / scopes. They are as follows:
- Scope 1: `CollateralFTokenAdaptor.sol` - Accounting of the collateral provided to Fraxlend, and the logic for collateral provision. 
- Scope 2: `DebtFTokenAdaptor.sol` - Allows debt positions to be taken from FraxLend. 
- Additional Scope: Incorporate the sfrxETH pricing extension w/ `PriceRouter`

### Scope 1: `CollateralFTokenAdaptor.sol`
- [x] Stub out code for reading the balance of collateral provided to Fraxlend contracts.
- [x] Write rough draft of adaptor that simply monitors the collateral provided from the cellar for a respective Fraxlend Pair.
- [x] Add strategist functions to deposit and remove collateral
- [x] Add base functions
- [ ] Write unit tests
- [ ] PR Review and refine
- [ ] Finalize

### Scope 2: `DebtFTokenAdaptor.sol` TODOs:

- [x] Stub out code w/ referenced docs on how to take debt while keeping track of health factor
- [x] Stub out strategist functions
- [x] Peer review w/ Crispy on adaptor design/approach
- [x] Revise implementation code
- [ ] Review revised helpers w/ Crispy --> specifically `_isSolvent()`
- [ ] Write unit tests and integration tests
- [ ] Peer review & Finalize
- [ ] Final Peer review & Finalize for audit submission
- [ ] Make github issues && finalize clarity on some questions from FraxLend if any

## Notes:
- See Aave AToken and debt adaptors as reference
- tl;dr - 2 adaptors, 1st one is for providing collateral, 2nd is for borrowing. All records are stored in `VaultAccount` within the respective Fraxlend Pair contracts.
- From looking at recent github repo for FraxLend, the borrow function only checks that the user's account doesn't go below 0 (totalCollat - totalBorrow). It does not check for health factor. Thus we have to check before allowing a borrow to transact. 

_From Crispy for CollateralAdaptor:_
- For base function implementation code in `CollateralFTokenAdaptor`, could add logic to the deposit and withdraw --> withdraw might be not great.
   - could check out the AaveV2Morpho if you want simpler, the ATokenAdaptor is the way we would want to go... --> Strategists do like that they can have some liquid asset in this. 

## Live Discussion Topics (discussed during the Draft PR stage)

_`isSolvent()` Helper_
- This is VIP aspect to design. Are we using an internal helper from the adaptor, or somehow accessing the return value from the FraxlendPair contract (not it is not external or public)?
TODO: See helper `_isSolvent` in CollateralFTokenAdaptor and DebtFTokenAdaptor --> NOTE that it is the same helper in CollateralFTokenAdaptor, so once either is figured out just copy and paste the proper solution.

## References:
1. Fraxlend Docs: https://docs.frax.finance/fraxlend/lending & https://docs.frax.finance/fraxlend/advanced-concepts/vault-account specifically.

---

<details markdown='1'><summary> ℹ️ Extra Notes </summary>

</details>

<!--- example - Design decisions made in PR>
<!--- example - Remember running `forge fmt` before committing the PR>
<!--- example - Reference for how to run relevant tests i.e. forge test --match-contract X --fork-url $RPC_URL -vvv>

### CODE REVIEW

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    /**
+     * @notice Returns the cellar's balance of the collateralAsset position.
+     * @param adaptorData the collateral asset deposited into Fraxlend
+     * NOTE: CRISPY QUESTION - TODO: confirm that this works... EIN - FraxlendCore doesn't have an interface to access the getter. So we'll have to have it here.
+     * TODO: confirm that there is no need to typeCast on the decoded adaptorData when needing address, ERC20 or vice versa.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        ICollateralFToken fToken = ICollateralFToken(abi.decode(adaptorData, (address))); // TODO: CRISPY QUESTION - could change adaptorData to be collateralToken but then it won't be consistent with rest of fraxlend adaptors.
```
Comment (crispymangoes) [inline src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:104]: It is 100% fine for this adaptor data to be different from the other fraxlend adaptors

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    //============================================ Strategist Functions ===========================================
+
+    /**
+     * @notice Allows strategists to add collateral to the respective cellar position on FraxLend, enabling borrowing.
+     * @dev `borrowTokens()` within DebtFTokenAdaptor can increase cellar collateralPosition too. `borrowTokens()` cannot be used without a trusted CollateralFTokenAdaptor position in the same cellar because CollateralFTokenAdaptors track the collateral within a Fraxlend pair. CRISPY QUESTION - See TODO: within DebtFTokenAdaptor regarding this.
+     */
+    function addCollateral(
+        ERC20 _collateralToken,
+        uint256 _collateralToDeposit,
+        IFToken _fToken
```
Comment (crispymangoes) [inline src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:132]: I think adaptordata should contain the fToken address, and the collateral token. Then we can verify this position is setup in the cellar, and that the fToken address is a real fToken address.

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    ) public {
+        // _validateFToken(collateralToken); // TODO: CRISPY QUESTION - we could have a validation helper but Fraxlend checks if there is even a collateral position for whomever is attempting to borrow. So it is not necessary unless we want to save Strategists gas during reversions. I have left this here in case I am wrong.
+        // amountToDeposit = _maxAvailable(collateralToken, amountToDeposit); // TODO: CRISPY QUESTION - not sure if we want to deposit the max or not by default... It depends if the strategist wants to do uint256.max or not for these cellars.
+        address fraxlendPair = address(_fToken);
+        _collateralToken.safeApprove(fraxlendPair, _collateralToDeposit);
+        _fraxLendPair.addCollateral(_collateralToDeposit, address(this));
+
+        // TODO: CRISPY QUESTION - should we check that the LTV is in a healthy state? Check that the FraxlendCore doesn't already do that.
```
Comment (crispymangoes) [inline src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:140]: This function is adding collateral, so I dont think we need to check the LTV, but on `removeCollateral` we should check the LTV.

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        IFToken fToken = abi.decode(adaptorData, (IFToken));
+        return _toBorrowAmount(fToken, fToken.userBorrowShares(msg.sender), false, ACCOUNT_FOR_INTEREST);
+    }
+
+    /**
+     * @notice Returns FRAX.
+     */
+    function assetOf(bytes memory) public view override returns (ERC20) {
+        return FRAX;
```
Comment (crispymangoes) [inline src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:133]: Could rename this to debtToken or something since in the future it could be something different

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+     * @param amountToRepay the amount of `tokenToRepay` to repay with.
+     * NOTE: this version of repayFraxlendDebt is meant to offer an alternative to doing it similar to the AaveDebtTokenAdaptor (see notes below)
+     * TODO: CRISPY --> from `FraxlendCore.sol` functions `repayAsset()` and `_repayAsset()` I don't see specific responses from the contracts if the amount of shares repaid >> amount of borrowShares even owed. Unless `-=` takes care of that.
+     * Assuming the above is true, then we'll need to check how much borrowShares we need to pay. Then we can get the total amount of Frax we need to repay. So we could specify the amount of FRAX we're willing to repay (could be max), and then we calculate the amount of FRAX owing from fraxlend, then we check the two values to make sure it is less than the amount we're willing to repay. From there we can simply repay the shares. We then check that we have the expected change in borrowShares as per FraxLend accounting.
+     */
+    function repayFraxlendDebt2(
```
Comment (crispymangoes) [inline src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:201]: I think this one is better. I want to add some new logic to all of our lending adaptors, repay max debt, which either caps it to the debt owed, or it is capped by the ERC20s the cellar has.

Like if 10k is owed, but the cellar only has 8k, the logic should have it repay 8k of debt

**Review by crispymangoes — CHANGES_REQUESTED** [review]
Lookin good my guy

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol (hunk 1 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:104]: Sounds good!

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol (hunk 2 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:132]: That makes sense. Sweet, thanks!

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+     * @notice Allows strategists to add collateral to the respective cellar position on FraxLend, enabling borrowing.
+     * @dev `borrowTokens()` within DebtFTokenAdaptor can increase cellar collateralPosition too. `borrowTokens()` cannot be used without a trusted CollateralFTokenAdaptor position in the same cellar because CollateralFTokenAdaptors track the collateral within a Fraxlend pair. CRISPY QUESTION - See TODO: within DebtFTokenAdaptor regarding this.
+     */
+    function addCollateral(
+        ERC20 _collateralToken,
+        uint256 _collateralToDeposit,
+        IFToken _fToken
+    ) public {
+        // _validateFToken(collateralToken); // TODO: CRISPY QUESTION - we could have a validation helper but Fraxlend checks if there is even a collateral position for whomever is attempting to borrow. So it is not necessary unless we want to save Strategists gas during reversions. I have left this here in case I am wrong.
+        // amountToDeposit = _maxAvailable(collateralToken, amountToDeposit); // TODO: CRISPY QUESTION - not sure if we want to deposit the max or not by default... It depends if the strategist wants to do uint256.max or not for these cellars.
```
Comment (0xEinCodes) [inline src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:135]: @crispymangoes did you have any thoughts on this?

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+     *         this position must return 0 for withdrawableFrom.
+     */
+    function withdrawableFrom(bytes memory, bytes memory) public pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Returns the cellar's balance of the respective Fraxlend debtToken calculated from cellar borrow shares
+     * @param adaptorData encoded fToken (fraxlendPair) for this position
+     * TODO: CRISPY QUESTION - should we call them fTokens, or fraxlendPairs? I've gone ahead and made them fTokens to stay consistent, but it could be argued that for lending/borrowing it makes more sense to specify it as FraxlendPair.
```
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:121]: Sounds good, I'll make the adjustment so they are called `fraxlendPairs`

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+     * TODO: CRISPY QUESTION - From looking at AaveDebtTokenAdaptor balanceOf() it looks like it just reports the amount of debtTokens it has, not the overall owed amount? As in, if the cellar borrowed a bunch of DAI from Aave, and then used it, it wouldn't account for it anymore. Need to think on this more. I'd think we need to show the amount of debt that this cellar has/is accruing (thus account for interest would be set to true).
```
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:122]: Okay, so I think my question was more about whether they used the `borrowAmount` or some portion of it to do some other activity. If that is done, then there are other trusted `positions` to account for the new assets involved. Alright, so it is all taken care of. The debt grows on the `fraxlendPair` side of things and thus `balanceOf()` querying the `fraxlendPair.userBorrowShares(msg.sender)` for the cellar is accurate.

tl;dr - I'll leave it as is then.

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+     * TODO: See helper `_isSolvent` --> NOTE that it is the same helper in CollateralFTokenAdaptor, so once either is figured out just copy and paste the proper solution.
+     */
+    function borrowFromFraxlend(IFToken fToken, uint256 amountToBorrow) public {
+        // Check that debt position is properly set up to be tracked in the Cellar.
+        bytes32 positionHash = keccak256(abi.encode(identifier(), true, abi.encode(address(fToken))));
+        uint32 positionId = Cellar(address(this)).registry().getPositionHashToPositionId(positionHash);
+        if (!Cellar(address(this)).isPositionUsed(positionId))
+            revert DebtFTokenAdaptor__FTokenPositionsMustBeTracked(address(fToken));
+
+        fToken.borrowAsset(amountToBorrow, 0, msg.sender); // TODO: CRISPY - Could have second param, `_collateralAmount` nonzero by having Strategist specify an amount if they want to top up their position or not.
```
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:161]: Agreed, it makes things more `explicit` in terms of whether calls will be 'positive' or 'debt' oriented moves from strategists.

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol (hunk 5 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:201]: kk, sounds good. I'll have this one be the live version of the function.

File: src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+    /**
+     * @notice Returns the cellar's balance of the collateralAsset position.
+     * @param adaptorData the collateral asset deposited into Fraxlend
+     * NOTE: CRISPY QUESTION - TODO: confirm that this works... EIN - FraxlendCore doesn't have an interface to access the getter. So we'll have to have it here.
+     * TODO: confirm that there is no need to typeCast on the decoded adaptorData when needing address, ERC20 or vice versa.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        ICollateralFToken fToken = ICollateralFToken(abi.decode(adaptorData, (address))); // TODO: CRISPY QUESTION - could change adaptorData to be collateralToken but then it won't be consistent with rest of fraxlend adaptors.
+        return fToken.userCollateralBalance(msg.sender) + fToken.collateralContract().balanceOf(msg.sender); // reports balance of collateral provided to protocol.
```
Comment (0xEinCodes) [inline src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol:105]: I didn't realize this but, we'll need `collateral` as a `erc20Adaptor` trusted position asset.  It is accounted then with that adaptor for any lingering `collateral` ERC20 balance for the cellar.

File: src/modules/adaptors/Frax/DebtFTokenAdaptor.sol (hunk 4 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Frax/DebtFTokenAdaptor.sol:133]: I think I made a comment on this but:

- Community from FRAX seems to think that they will not ever have a different `debtAsset`

That said, we'll make it the return var from `FraxlendPair.asset()`

**Comment by 0xEinCodes** [comment]
Closing this draft PR and continuing in PR #134. See description and comments there.
